RADIUSS packages update (Starting over #39613)

adrienbernede opened the pull request:
This PR updates BLT, Caliper, Camp, CHAI, RAJA and Umpire packages.

The update consists in adding newer versions and variants (or removing some), updating the package logic, and sharing part of the package implementation, either in as part of the CachedCMakePackage, or with utility functions in the BLT package.

Changes in the CI are limited to removing deleted variants.

Some patches were necessary to get the CI to pass.


Detailed update process follows.

### Step 1 (3a89273):
We add 2023.06 versions of Umpire and Camp and only make minor improvement changes.

### Step 2 (4233b17):
Spack won’t concretize e4s env with any umpire 2023.06, while it does so in #41145. The constraint must come for either blt or camp, which is also need by RAJA as a CHAI dependency. I am bumping RAJA version now.
However, some a CHAI spec with `~raja` does not build umpire@2023.06. Expecting this attempt to pass CI, but not build Umpire@2023.06 for unknown reason...

### Step 3 (4517fbc):
The problem in step one in solved by adding a commit hash to specify the 2023.06.0 version in Camp.
Now Umpire 2023.06.0 is concretized, but the build fails.

### Step 4 (bb0bbb1):
We add CHAI 2023.06 version (but with Umpire failing, that is not going to have any effect).

### Step 5 (1e44364):
We add missing conflicts to start the process of getting Umpire to build. However, the error appears more profound: In our CI we never built rocm support with gcc, we may need to change the compiler.
```
                  [...]
                  /tmp/root/spack-stage/spack-stage-umpire-2023.06.0-lhi6wk7lzsmyeuczujsnfyfs32jmri3y/spack-src/examples/allocator.c
  >> 354    gcc: error: unrecognized command-line option '--rocm-path=/opt/rocm
            -5.4.3'
  >> 355    gcc: error: unrecognized command-line option '-fgpu-rdc'
  >> 356    make[2]: *** [examples/CMakeFiles/allocator_c.dir/build.make:79: ex
            amples/CMakeFiles/allocator_c.dir/allocator.c.o] Error 1
```
We have two fallback solutions:
- In our local package device alloc is turned off by default.
- In our local package examples are turned off by default.

It turns out the problem is with the allocator.c example, which extension forces the use of gcc instead of hipcc. The flag `--rocm-path` being unknown to gcc, this example build fails.

### Step 6 (35347ff):
When Ascent is built with recent blt and umpire (typically blt@0.5.3 and umpire@2022.10), there is a conflict on targets exported by BLT in both Camp and Umpire. Both use BLT but Umpire use it to export more targets than Camp. The already exported target causes the crash.
~This could be solved by each project using Camp and Umpire in their build system (I don’t have a hand on that). Or it would require massive patching in BLT. Instead, the only viable solution appears to be waiting for the coming BLT, Umpire & Co. releases which will bring in a fix.~ WORKAROUND: see **Step 9 and 10**.

### Step 7 (c726cdf):
Adding a conflict between `device_alloc` and `~rocm~cuda` in Umpire resulted in the build of an older version of DRay and Ascent. While the cause for this is unexplained, the CI ends up building DRay@1.8.0 with RAJA@0.13.0, which should not be possible, but is not expressed well in DRay packages. We patch that for the issue we have.
~TODO: @cyrush Please have a look, a constraint like `depends_on("raja@:0.14", when="@0.1.7:")` is not preventing a recent DRay to build with a very old RAJA.~ DONE: vetted by @cyrush (see https://github.com/spack/spack/pull/41375#issuecomment-1839186718)

### Step 8 (9e90134):
Appearing at step 5, we work around the issue with allocator.c example by disabling the examples by default. @davidbeckingsale I think it would be nice to build the examples, but we need a way to restrict rocm flags to rocm compilers.
TODO: We can reconsider this when the rocm management has been updated. So far it hasn’t in this PR.

### Step 9 (5f2714f 39b7a11):
Since CI is passing, we move forward with adding new variants to RAJA and Umpire. The CI then fails because jobs that hadn’t be rebuilt so far are now rebuilt, revealing incomplete coverage of **Step 6**.
~TODO: We can either enforce umpire@2022.03 for every ascent spec, or apply @cyrush patch (see https://github.com/spack/spack/pull/41375#issuecomment-1839186718)~ **DONE: Step 10**

### Step 10 (77e2457):
We apply @cyrush patch to umpire. This fix is not perfect as it is not addressing the root cause. We will have to restrict the patch usage to `2022.10.0:2023.06` once the BLT fix is integrated to Umpire. (104c30b, a653893, c5460a2, 7c6bff2 and cc1eae6 are fixes for the consequences of this change, implying adding corresponding variants to CHAI).

### Step 11 (a21e5a9):
At **Step 9** we added werror which causes errors in some build configurations, in both umpire@6.0 and umpire@2023.06, @davidbeckingsale you may want to have a look. Werror is turned of to ignore these.

### Step 12 (39b8c22 ... cbf5357):
With the CI passing again, we further the effort to sync with #39613, except for the enforcement of umpire@2022.03 in ascent which we removed using @cyrush patch.

### Many more steps...

adrienbernede committed:
Starting over

adrienbernede committed:
Add 2023.06 to RAJA, hoping to allow the build of Umpire@2023.06 via CHAI

adrienbernede committed:
Adding commit for 2023.06.0 version of Camp hoping to get it built by Spack CI

adrienbernede committed:
Adding CHAI 2022.10 and 2023.06

adrienbernede committed:
Adding new device_alloc conflicts

adrienbernede committed:
Enforcing Umpire version for Ascent: know issue solved in next release batch

adrienbernede committed:
Patch loosely defined umpire dependency in dray, still not great but fixes CI

adrienbernede committed:
We are missing a way to not specify --rocm-path for .c files which are not compiled with hipcc

adrienbernede committed:
Move helper functions to BLT

adrienbernede committed:
Fix style

adrienbernede committed:
Fix missing function: logic moved to Cached CMake package

cyrush commented:
I think Step 6 is a known issue with BLT vs Umpire vs Camp.

I actually tested a patch for umpire vs ascent that avoids using the BLT OpenMP logic and instead uses the bona fide CMake targets:

https://github.com/Alpine-DAV/ascent/pull/1198

I think the global path forward is to use a new BLT, but my patch provides an alternate solution.  


Step 7: Devil Ray is no longer a stand alone project, it is developed as part of Ascent. I am not sure there is much value in trying to maintain all these combos with older Devil Ray. 

I am fine with any constraints that keep noise related to Dray builds down. Your additions are good b/c Dray should not be getting built in new versions regardless.

adrienbernede committed:
Add variants to RAJA

adrienbernede committed:
Add variants to Umpire

adrienbernede committed:
Fix typo

adrienbernede committed:
Add patch to Umpire to work around BLT dual import in Ascent

adrienbernede committed:
Propagate Umpire patch effect to CHAI package

adrienbernede committed:
Add variants (and MPI support) to CHAI

adrienbernede committed:
Remove benchmarks and tests variants form CHAI specs in CI: new variants and defaults

adrienbernede committed:
Remove CHAI benchmarks variant in CARE

adrienbernede committed:
Fix syntax

adrienbernede committed:
Umpire: turn off werror by default

adrienbernede committed:
Remove duplicate logic in camp package

adrienbernede committed:
Further sync with original branch (local radiuss packages)

adrienbernede committed:
Remove umpire requirement in spack CI, patching works

adrienbernede committed:
Remove duplicate logic in raja package

adrienbernede committed:
Cosmetic change

adrienbernede committed:
Attempt at moving cuda and rocm logic to CachedCMakePackage. Only change is removing HIP_CXX_COMPILER in favor of CMAKE_CXX_COMPILER

adrienbernede committed:
Fix syntax

adrienbernede committed:
Fix style

adrienbernede committed:
Only define HIP_HIPCC_FLAGS in packages to avoid collisions with CachedCMakePackage definition

adrienbernede committed:
Fix oversights

adrienbernede committed:
Limit changes with respect to develop

adrienbernede committed:
Fix style

adrienbernede commented:
@spackbot rebuild everything

adrienbernede committed:
Uncomment logic that is present in develop (but in camp package)

adrienbernede committed:
Revert previous logic, set hipcc as main CXX compiler

adrienbernede committed:
Clean CHAI package (remove logic shared in CachedCMakePackage)

adrienbernede committed:
clean unneeded hipcc flags

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update-start-over

adrienbernede committed:
Style fix

adrienbernede committed:
GitLab CI: Remove benchmark variant in CHAI spec

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update-start-over

adrienbernede commented:
@tldahlgren I finally got the CI all green in this PR, which is a rewrite of #39613. There are some other tests failing (coverage, and others). May I get guidance about dealing with those ? Thanks

adrienbernede commented:
@spackbot rebuild everything

adrienbernede commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -252,6 +259,13 @@ def initconfig_hardware_entries(self):
                 entries.append(
                     cmake_cache_string("CMAKE_CUDA_ARCHITECTURES", "{0}".format(arch_str))
                 )
+                # TODO: Additional definitions that may not be needed anymore
+                # This is an imperfect merge of the radiuss way and the "spack"
+                # way: radiuss used to define only the first arch in the list,
+                # even for CMAKE_CUDA_ARCHITECTURE. What do we want?
+                entries.append(cmake_cache_string("CUDA_ARCH", "sm_{0}".format(archs[0])))
+                flag = "-arch sm_{0}".format(archs[0])
+                entries.append(cmake_cache_string("CMAKE_CUDA_FLAGS", "{0}".format(flag)))
```
@davidbeckingsale @rhornung67

adrienbernede commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -260,9 +274,18 @@ def initconfig_hardware_entries(self):
-            entries.append(
-                cmake_cache_path("HIP_CXX_COMPILER", "{0}".format(self.spec["hip"].hipcc))
-            )
+
+            # entries.append(
+            #     cmake_cache_path("HIP_CXX_COMPILER", "{0}".format(self.spec["hip"].hipcc))
+            # )
+
+            # entries.append(
+            #     cmake_cache_path(
+            #         "HIP_CLANG_INCLUDE_PATH",
+            #         glob.glob("{}/lib/clang/*/include".format(spec["llvm-amdgpu"].prefix))[0],
+            #     )
+            # )
+
```
```suggestion

```

adrienbernede commented on var/spack/repos/builtin/packages/blt/package.py:
```diff
@@ -3,9 +3,90 @@
+
+        version = "{0}".format(compiler.version)
+
+        if version == "16.0.0":
```
```suggestion
        if version == "16.0.0" or version == "16.0.1":
```

adrienbernede commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -14,6 +14,8 @@
 
 from .cmake import CMakeBuilder, CMakePackage
 
+# import glob
+
```
```suggestion
```

adrienbernede commented on var/spack/repos/builtin/packages/camp/package.py:
```diff
@@ -68,33 +55,27 @@ def cmake_args(self):
-            hip_repair_options(options, spec)
+            options.append("-DHIP_ROOT_DIR={0}".format(spec["hip"].prefix))
 
             archs = self.spec.variants["amdgpu_target"].value
             if archs != "none":
                 arch_str = ",".join(archs)
                 options.append("-DHIP_HIPCC_FLAGS=--amdgpu-target={0}".format(arch_str))
```
Should not be necessary anymore.

adrienbernede commented on var/spack/repos/builtin/packages/raja/package.py:
```diff
@@ -246,19 +308,50 @@ def initconfig_package_entries(self):
             entries.append(cmake_cache_option("ENABLE_TESTS", False))
         else:
             entries.append(cmake_cache_option("ENABLE_TESTS", True))
+            if "+run-all-tests" not in spec:
+                if spec.satisfies("%clang@12.0.0:13.9.999"):
+                    entries.append(
+                        cmake_cache_string(
+                            "CTEST_CUSTOM_TESTS_IGNORE",
+                            "test-algorithm-sort-OpenMP.exe;test-algorithm-stable-sort-OpenMP.exe",
+                        )
+                    )
+                excluded_tests = [
+                    "test-algorithm-sort-Cuda.exe",
+                    "test-algorithm-stable-sort-Cuda.exe",
+                    "test-algorithm-sort-OpenMP.exe",
+                    "test-algorithm-stable-sort-OpenMP.exe",
+                ]
+                if spec.satisfies("+cuda %clang@12.0.0:13.9.999"):
+                    entries.append(
+                        cmake_cache_string("CTEST_CUSTOM_TESTS_IGNORE", ";".join(excluded_tests))
+                    )
+                if spec.satisfies("+cuda %xl@16.1.1.12"):
+                    entries.append(
+                        cmake_cache_string(
+                            "CTEST_CUSTOM_TESTS_IGNORE",
+                            "test-algorithm-sort-Cuda.exe;test-algorithm-stable-sort-Cuda.exe",
+                        )
+                    )
```
@rhornung67 I’m not sure why this is considered pure new addition. Maybe it was removed from RAJA package in Spack upstream ? I don’t think it was specific to our local version. I need to check the history.

adrienbernede replied:
This is likely a new addition :)

adrienbernede reviewed (commented):
Self-reviewing, ’cause why not ?

adrienbernede committed:
Apply suggestions from code review

adrienbernede commented:
@tldahlgren I asked spack bot for a full rebuild because this PR touches the Cached CMake Package and changes there can impact packages using this build system while they are not detected as a change in the recipe. I’d rather be extra safe: I’ve seen new error pop up that way working on these PRs.

adrienbernede committed:
Fix style and hip arch spec in camp

adrienbernede committed:
Import minor changes from radiuss-spack-configs

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update-start-over

adrienbernede commented:
@tldahlgren what’s with code coverage, what can I do to get them to pass ?

adrienbernede commented:
@tldahlgren @trws @davidbeckingsale I’d like your review on this.
The only remaining failure is also present in develop

adrienbernede committed:
Merge tag 'develop-2024-01-21' into llnl/radiuss-packages-update-start-over

Snapshot release 2024-01-21

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update-start-over

adrienbernede committed:
Simplify syntax

adrienbernede commented:
@tldahlgren @trws should I ping someone else ? Conflicts are often painful to deal with in a "large" PR... I’d rather do the work now to get this merged. Thank you.

adrienbernede commented:
@scheibelp I’m looking for feedback on this. I’m hoping to get this merged before other conflicting changes show up.

trws commented:
I'm looking it over, but right at the start running into this ensuring that we don't use the spack wrappers for hip code.  Is that intentional?

trws commented on var/spack/repos/builtin/packages/blt/package.py:
```diff
@@ -3,9 +3,90 @@
+            if flags:
+                options.append(cmake_cache_string("BLT_EXE_LINKER_FLAGS", flags, description))
+
+            excluded_root = "/usr/tce/packages/gcc/gcc-4.9.3"
```
Why is this a completely llnl-specific path in a public spack package?

adrienbernede replied:
Well, the point of this PR is to share what makes our packages build on LC systems. I agree that ideally there would be no need for machine specific stuff. Our vision is that as long as it does not hurt others, then it is preferable to have those requirement right in upstream Spack.
We have been maintaining a local version of those packages, and it’s been a maintenance hassle with impact on Spack itself because LLNL packages in Spack used to be unlikely to build on LC systems...
What is your take on this?

trws replied:
My take on this is that the spack package should work on an LC system and at a minimum a stock rhel system of a similar version.  This specific rpathing code shouldn't be necessary, especially after the recent gcc-runtime changes, so after a merge or rebase it might be kinda moot after a test, but if we merge something like this, I'd like to at least see an issue with what needs to happen to make it portable.  In this specific case if the `gcc-runtime` package injection and rpathing doesn't fix it, which I think it will, then it would be fixed by fixing #40848 as discussed in this [thread on maintainers slack](https://spackpm.slack.com/archives/G7XED8NAD/p1706105393206209) where issues have arisen due to either not getting RPATHs to system compilers that need them or adding rpaths to system compilers with a sysroot that must not have them.  (CC: @haampie, @tgamblin, @scheibelp pulling you in from the slack thread)

adrienbernede replied:
I’m fine with testing whether this is still necessary.

adrienbernede replied:
Here is what happens building Umpire on lassen if I remove everything in the `blt_link_helpers` function:

(this is for build with clang_14_0_5 using gcc_8_3_1 toolchain (set in the compilers flags) and cuda_11_7_0).

```
[ 13%] Linking CXX executable ../../../tests/blt_cuda_gtest_smoke
cd /dev/shm/lassen589-1660612/build_lassen-blueos_3_ppc64le_ib_p9-clang@14.0.5.gcc.8.3.1-xwyxixtw/blt/tests/smoke && /usr/tce/packages/cmake/cmake-3.23.1/bin/cmake -E cmake_link_script CMakeFiles/blt_cuda_gtest_smoke.dir/link.txt --verbose=1
/usr/tce/packages/clang/clang-14.0.5/bin/clang++  -Wall -Wextra  --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1     -Werror  -O3 -DNDEBUG -L/usr/tce/packages/cuda/cuda-11.7.0/lib64 CMakeFiles/blt_cuda_gtest_smoke.dir/blt_cuda_gtest_smoke.cpp.o -o ../../../tests/blt_cuda_gtest_smoke   -L/usr/tce/packages/cuda/cuda-11.7.0/nvidia/targets/ppc64le-linux/lib/stubs  -L/usr/tce/packages/cuda/cuda-11.7.0/nvidia/targets/ppc64le-linux/lib  -L/usr/tce/packages/gcc/gcc-4.9.3/lib64/gcc/powerpc64le-unknown-linux-gnu/4.9.3  -L/usr/tce/packages/gcc/gcc-4.9.3/lib64  -Wl,-rpath,/dev/shm/lassen589-1660612/clang-14.0.5.gcc.8.3.1/umpire-develop-xwyxixtwe7zh4eqqp7mc4spm7if7fagj/lib:/dev/shm/lassen589-1660612/clang-14.0.5.gcc.8.3.1/umpire-develop-xwyxixtwe7zh4eqqp7mc4spm7if7fagj/lib64:/dev/shm/lassen589-1660612/clang-14.0.5.gcc.8.3.1/camp-main-vfmvkfia3v2hm5sdtuum4tci5fbi75he/lib:/usr/tce/packages/cuda/cuda-11.7.0/lib64:/dev/shm/lassen589-1660612/clang-14.0.5.gcc.8.3.1/fmt-10.2.1-l3zyrwzbxv5ria5h6oicoytjj66cjpp6/lib64:/usr/tce/packages/gcc/gcc-4.9.3/lib64:/usr/tce/packages/clang/clang-14.0.5/release/lib ../../../lib/libgtest_main.a ../../../lib/libgtest.a -lpthread /usr/tce/packages/cuda/cuda-11.7.0/lib64/libcudart_static.a -ldl /usr/lib64/librt.so -lpthread -lcudadevrt -lcudart_static -lrt -lpthread -ldl 
../../../lib/libgtest.a(gtest-all.cc.o): In function `testing::internal::GoogleTestFailureException::GoogleTestFailureException(testing::TestPartResult const&)':
gtest-all.cc:(.text+0x22e3c): undefined reference to `std::runtime_error::runtime_error(char const*)'
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[2]: *** [blt/tests/smoke/CMakeFiles/blt_cuda_gtest_smoke.dir/build.make:101: tests/blt_cuda_gtest_smoke] Error 1
gmake[2]: Leaving directory '/dev/shm/lassen589-1660612/build_lassen-blueos_3_ppc64le_ib_p9-clang@14.0.5.gcc.8.3.1-xwyxixtw'
gmake[1]: Leaving directory '/dev/shm/lassen589-1660612/build_lassen-blueos_3_ppc64le_ib_p9-clang@14.0.5.gcc.8.3.1-xwyxixtw'
gmake[1]: *** [CMakeFiles/Makefile2:1946: blt/tests/smoke/CMakeFiles/blt_cuda_gtest_smoke.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

Things like `-L/usr/tce/packages/gcc/gcc-4.9.3/lib64/gcc/powerpc64le-unknown-linux-gnu/4.9.3` result in the wrong stdc++ being used, therefore inducing the build error.

adrienbernede replied:
Please also note that Axom is already doing something similar here:
https://github.com/spack/spack/blob/22ad28bb17820eed041731814950531e24d52a30/var/spack/repos/builtin/packages/axom/package.py#L393

Which gives me an occasion to politely invoke @white238.

adrienbernede replied:
While the shared logic for hip and cuda has been integrated to the CachedCMakePackage (discussing the topic with Spack folks internally, this was identified as the better choice), I found that this LLNL specific logic, relying on BLT, would be better defined as an helper function in the BLT package.

trws replied:
Oh I see, this is a workaround for not passing the right flags somewhere.  There isn't a missing rpath, one of the cmake detection passes is finding the wrong GCC installation because the `--gcc-toolchain` flag is missing.  Without seeing your `CMakeFiles/CMakeOutput.log` file I can't say which one, but doing the same experiment with the same clang, gcc and nvcc on lassen works with this cmake configure line works just fine: `CXX=clang++ CC=clang cmake -B build -DCMAKE_C_FLAGS=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/ -DCMAKE_CUDA_FLAGS="-Xcompiler=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/" -DCMAKE_CXX_FLAGS=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/ -DCMAKE_CXX_COMPILER=clang++ -DENABLE_CUDA=On -DCMAKE_CUDA_ARCHITECTURES=70 -DCMAKE_EXE_LINKER_FLAGS="--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/" .`

I'm guessing the `CMAKE_CUDA_FLAGS` bit is missing for umpire somehow, based on the host_configs, but could be wrong.  The issue in the end is since the wrong gcc is found by cmake running either a compile or link phase for a language without the gcc-toolchain, the wrong implicit link path is added, so a wrong -L is added and it actually links with the wrong libstdc++.  Would you have an issue with adding the flags where appropriate, maybe even always in BLT based on the `spec_using_toolchain`? That way it should work everywhere that has separate gcc toolchains with separate libstdc++ instances.

adrienbernede replied:
OK, it looks like the cuda flag for the toolchain is missing indeed. I can try adding that.

Regarding the rpath for the missing stdc++ with fortran, I see it in axom too. (Although it looks like the implementation is more complete there.)
https://github.com/spack/spack/blob/22ad28bb17820eed041731814950531e24d52a30/var/spack/repos/builtin/packages/axom/package.py#L225

trws replied:
Oh interesting, I missed the fortran part, and didn't test a fortran target, maybe the rpath is separately needed for that.  The gcc-runtime package should be adding an rpath for the gcc version in use, but maybe we're not getting it because we haven't expressed to spack that we're using that gcc through clang.  @haampie is there a way to specify which gcc toolchain version is in use when not using `%gcc`?

adrienbernede replied:
@trws there was indeed an oversight regarding the gcc-toolchain flags in the cuda build command. It is now fixed.
This oversight does not fix the handling of gcc-toolchain however: we have the correct flag in #39613, and locally in our builds, which verifies in the scenario described above
```
-- CUDA Compile Flags:             -Xcompiler --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1
```
(see https://lc.llnl.gov/gitlab/radiuss/Umpire/-/jobs/1660612).

adrienbernede replied:
May I suggest that since we pass --gcc-toolchain=... through the compiler flags (in compilers.yaml), we may not end up with the exact same configuration you tried ?

I am going to print out the resulting hostconfig generated by Spack, as well as the CMakeFiles/CMakeOutput.log file.

trws replied:
I don't see it on the linker flags in there, which may be what's left. 
 I know this is nuts, there should be a better way to ensure this, but it will be more reliable if cmake is using the right gcc toolchain all the way down.  If that doesn't do it, go for it in the compiler flags or do what you need to do and we can break something out into a new issue to fix that part up rather than keeping this stuck.

adrienbernede replied:
I agree it is not in the linker flags, I came to the same conclusion later today comparing your build line with my logs.

I am currently looking at that, that could be it. Let's follow up on Monday.

trws replied:
Sounds good, sorry for my slow followup by the way, I'm out of town in all-day meetings this week so I only manage to look in sporadically. Will be back in town Monday so if you want to chat that should be a lot more feasible then.

adrienbernede replied:
I see no improvement with the appropriate CMAKE_EXE_LINKER_FLAGS:

```
-- CMAKE_C_FLAGS flags are:   -Wall -Wextra  --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1  -Werror 
-- CMAKE_CXX_FLAGS flags are:   -Wall -Wextra  --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1     -Werror 
-- CMAKE_Fortran_FLAGS flags are:  --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1
-- CMAKE_EXE_LINKER_FLAGS flags are:   --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 -Wl,-rpath,/usr/tce/packages/xl/xl-2021.09.22/lib
```

However, I now think I may have to apply the same to the CUDA linker flags:

```
-- CUDA Compile Flags:             -Xcompiler --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1
-- CUDA Link Flags:                -L/usr/tce/packages/cuda/cuda-11.2.0/lib64
-- BLT CUDA Separable Compilation: OFF
-- CUDA Separable Compilation:     
-- CUDA Link with NVCC:            
-- CUDA Implicit Link Libraries:   stdc++;m;gcc_s;gcc;c;gcc_s;gcc
-- CUDA Implicit Link Directories: /usr/tce/packages/cuda/cuda-11.2.0/nvidia/targets/ppc64le-linux/lib/stubs;/usr/tce/packages/cuda/cuda-11.2.0/nvidia/targets/ppc64le-linux/lib;/usr/tce/packages/gcc/gcc-4.9.3/lib64/gcc/powerpc64le-unknown-linux-gnu/4.9.3;/usr/tce/packages/gcc/gcc-4.9.3/lib64;/lib64;/usr/lib64;/usr/tce/packages/clang/clang-12.0.1/release/lib;/lib;/usr/lib
```

CUDA Implicit Link Directories still has some gcc 4.9.3 paths.

white238 replied:
CMake implicitly finds paths you do not want on BlueOS. We (BLT) had to go in after it found them and strip them away because I could not find anyway to prevent it by a sane method. 

Here is how we set the variable in Axom:

https://github.com/LLNL/axom/blob/f77fd2e5f260abec7db6b7577f0482452af74e57/scripts/spack/packages/axom/package.py#L393-L410

Here is how BLT removes the paths:

https://github.com/LLNL/blt/blob/148c53ecc8bcaad5eaa4c1e39cb8144b8f1388ae/cmake/SetupCompilerOptions.cmake#L361-L375

We also have the ability to remove libraries as well:

https://github.com/LLNL/blt/blob/148c53ecc8bcaad5eaa4c1e39cb8144b8f1388ae/cmake/SetupCompilerOptions.cmake#L380-L394

trws replied:
Right, I get that.  It's also kinda cool that BLT has that backported, since cmake now supports it since 3.27 it's great to have that available.

That said, I tested it and found a working solution that will work with the host toolchain in any location rather than looking for hardcoded paths a few comments up.  If we're missing one of the flags to get the `--gcc-toolchain` in all the right places then let's fix that rather than letting cmake explicitly link against the wrong library directories and hoping it happens to work at runtime with an rpath to something else.

Without being able to see the logs from one of the ones not working, I can't tell what isn't getting through.  If you want to take a look for yourself, open up `<build-dir>/CMakeFiles/CMakeOutput.log` and search for `Found candidate GCC installation`.  There will be somewhere between four and six of them, detecting the ABI and directories for C, CXX, linkers, and attendant versions for use with cuda if configured.  They should _all_ mention only the configured gcc-toolchain directory, if any of them mention the system default compiler we're missing telling it something and cmake is _explicitly forcing `-L` flags for incorrect directories_.  

After further testing, this is a refined version of the cmake invocation above I just tested with raja on lassen and rzansel with no issues, no system compiler sneaks in:
`cmake -B build -DCMAKE_C_FLAGS=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/ -DCMAKE_CUDA_FLAGS="-Xcompiler --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/" -DCMAKE_CXX_FLAGS=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/ -DCMAKE_EXE_LINKER_FLAGS="--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/" -DCMAKE_CXX_COMPILER=clang++ DCMAKE_C_COMPILER=clang -DENABLE_CUDA=On -DCMAKE_CUDA_ARCHITECTURES=70  .`

@adrienbernede, if you can email or give me the cmake log I might be able to say what's sneaking in.  Every single flag on that line is necessary, up until the `ENABLE_CUDA`, including the C compiler since that pulls in paths too if we enable tests or certain other things.

trws commented on var/spack/repos/builtin/packages/blt/package.py:
```diff
@@ -3,9 +3,90 @@
+    if "cce" in compiler.cxx:
+        description = "Adds a missing rpath for libraries " "associated with the fortran compiler"
+        # Here is where to find libs that work for fortran
+        libdir = "/opt/cray/pe/cce/{0}/cce-clang/x86_64/lib".format(compiler.version)
```
Same.

trws commented on var/spack/repos/builtin/packages/raja/package.py:
```diff
@@ -246,19 +308,50 @@ def initconfig_package_entries(self):
         else:
             entries.append(cmake_cache_option("ENABLE_EXERCISES", "+exercises" in spec))
 
+        # TODO: Treat the workaround when building tests with spack wrapper
+        #       For now, removing it to test CI, which builds tests outside of wrapper.
```
Is this still an issue?

adrienbernede replied:
I’ll have to look for the origin of this. Also, I never build the tests with spack wrapper. This may simply need to a clean-up.

trws replied:
The part below seems stale, since the wrappers remove -Werror by default.  If the TODO can go too that would be great, either way want to know what's going on.

trws commented:
Oh and the MacOS failure, that may just go away after an update but if not we'll need to fix that too.

scheibelp commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -132,6 +132,11 @@ def initconfig_compiler_entries(self):
 
+        # We defined hipcc as top-level compiler for packages when +rocm.
+        # This avoid problems coming from rocm flags being applied to another compiler.
+        if "+rocm" in spec:
+            entries.insert(0, cmake_cache_path("CMAKE_CXX_COMPILER", self.spec["hip"].hipcc))
```
I'm curious what goes wrong when this is removed. I'm also not clear what you mean with 

> This avoid problems coming from rocm flags being applied to another compiler.

do you mean this isn't needed for some cases of `spec%compiler` but is needed for others? Or is this because you have HIP-specific compiler flags (e.g. in your `compilers.yaml`)?

adrienbernede replied:
This was already in Spack, both below in the `CachedCMakePackage` but through the `HIP_CXX_COMPILER`, which I removed, and in the packages, using the `CMAKE_CXX_COMPILER`.

Now, I agree this shouldn’t be necessary, but it is, especially in Spack CI. Hip code should be built with `rocmcc`. Using hipcc is, as I understand it, a deprecated practice. However, Spack CI is still asking for `+rocm` builds with `%gcc`.

I simplified my PR compared to #39613 because I was both making changes in the packages and trying to update the handling of the hip / rocm. I chose to keep things about the same in the present PR, and submit a subsequent one.

For that subsequent PR, @davidbeckingsale suggested that we have a discussion with the appropriate Spack audience regarding precisely this issue.

adrienbernede replied:
A simple example of what goes wrong without this is that `gcc` will not recognize `--rocm-path=...`.

trws replied:
Right, but hipcc shouldn't be set as the CMAKE_CXX_COMPILER, but the CMAKE_HIP_COMPILER unless something in spack is accidentally passing a flag to CXX_FLAGS that doesn't belong there.

davidbeckingsale replied:
This is required after switching to the native target-based support hip in BLT. This means a HIP capable compiler must be used, but as Adrien says, Spack CI builds things that are +rocm%gcc, which really doesn't make sense since HIP has no notion of host compiler in the way cuda does.

adrienbernede replied:
Thanks @davidbeckingsale. I remember trying CMAKE_HIP_COMPILER, which is actually what I have in #39613. But the tests there are failing with things like.
```
  >> 128    g++: error: unrecognized command-line option '--rocm-path=/opt/rocm
            -5.4.3'
```
I could re-run the same test here, but I’d like to emphasize that I am only doing what was already there before this PR:
https://github.com/spack/spack/blob/22ad28bb17820eed041731814950531e24d52a30/var/spack/repos/builtin/packages/raja/package.py#L196

The choice here is not to solve everything, trying that got me into too many problems at the same time, but rather to make some progress, incrementally. Hopefully gathering the logic in a single location is seen as a form of progress ;).

trws replied:
I'm probably missing something, but wouldn't that mean `CMAKE_CXX_COMPILER=g++` and `CMAKE_HIP_COMPILER=amdclang++`?  Using `hipcc` doesn't seem to make that better so much as more confusing.  If we end up having to leave this and fix it in another issue, I can probably live with it, but it seems pretty unfortunate that all non-hip code is being compiled with an entirely different compiler than spack has been asked to use, rather than just all host code _in hip compilation units_.

scheibelp replied:
> This was already in Spack, both below in the CachedCMakePackage but through the HIP_CXX_COMPILER, which I removed, and in the packages, using the CMAKE_CXX_COMPILER

I thought HIP_CXX_COMPILER and CMAKE_CXX_COMPILER were different: the former is used specifically for hip-enabled codes (which might not be all the code in the project).

Apologies if I'm rehashing what's already been stated.

>  g++: error: unrecognized command-line option '--rocm-path=/opt/rocm-5.4.3'

If I understand, you have explicit cflags in your config somewhere. If I'm right, most CMake packages set ROCM_PATH (at which point CMake knows how to invoke each compiler with the proper args).

adrienbernede committed:
Merge commit '5034919d23' into llnl/radiuss-packages-update-start-over

adrienbernede committed:
Apply gcc-toolchain flag to cuda compile command

adrienbernede committed:
Merge tag 'develop-2024-03-03' into llnl/radiuss-packages-update-start-over

Snapshot release 2024-03-03

pranav-sivaraman commented:
I've been testing this PR and also trying the packages from  LLNL/radiuss-spack-configs. For some reason when I trying enabling `+desul` I can get an error building RAJA saying that the `tpl/desul` directory is empty. This is probably because its not fetching the git submodules when building RAJA. One potential fix is to fetch the tarballs for RAJA, but I am not sure what the proper fix is.

adrienbernede commented:
@pranav-sivaraman thank you for trying this PR and reporting issues. Indeed, I don’t think we are doing a full build RAJA with desul atomics _through spack_ anywhere. I’ll try that tomorrow and see if I can reproduce your issue.

For the records, this PR will start moving forward again soon, we have been waiting for changes on our systems that should result in simplified logic in our packages, and clear the reviewers concerns.

rhornung67 commented:
> @pranav-sivaraman thank you for trying this PR and reporting issues. Indeed, I don’t think we are doing a full build RAJA with desul atomics _through spack_ anywhere. I’ll try that tomorrow and see if I can reproduce your issue.

@adrienbernede we have a GitLab build and test pipeline for RAJA using desul atomics on each of the 5 LC platforms we regularly test. Are you referring to a build with external desul vs. desul submodule?

trws commented:
> > @pranav-sivaraman thank you for trying this PR and reporting issues. Indeed, I don’t think we are doing a full build RAJA with desul atomics _through spack_ anywhere. I’ll try that tomorrow and see if I can reproduce your issue.
> 
> @adrienbernede we have a GitLab build and test pipeline for RAJA using desul atomics on each of the 5 LC platforms we regularly test. Are you referring to a build with external desul vs. desul submodule?

The _through spack_ part of that is what's operative @rhornung67, there's likely an issue with the spack package handling of desul with raja, or so it seems.

adrienbernede commented:
@rhornung67, as @trws stated, I purposefully highlighted "through spack" since even if we do have "+desul" targets in the CI, we do not actually build RAJA itself through Spack in the CI (we build the dependencies and generate the CMake config file).

adrienbernede committed:
Merge tag 'develop-2024-03-10' into llnl/radiuss-packages-update-start-over

Snapshot release 2024-03-10

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update-start-over

adrienbernede committed:
Update RADIUSS packages with 2024.02 release changes

adrienbernede committed:
Define tpl/desul as the only submodule to fetch

adrienbernede committed:
Work around an error with simplified logic in RAJA submodules management

adrienbernede commented:
@pranav-sivaraman would you mind trying again ?

@rhornung67 issue was that by preventing Spack from fetching any submodules, desul was found nowhere. I defined "tpl/desul" as the only submodule to fetch.

adrienbernede committed:
Fix style

adrienbernede committed:
Fix Umpire: limit version span for patch usage

adrienbernede committed:
Fix CHAI: limit version span for patch usage

adrienbernede committed:
Fix CHAI: missing dependency to BLT when @2024.02.0

adrienbernede committed:
Fix Axom: 0.8.2 does not support Umpire/RAJA @2024.02

pranav-sivaraman commented:
Just got back to this. I was able to successfully build RAJA with `+desul`.

adrienbernede committed:
Remove special logic in spack packages to enforce gcc toolchain

adrienbernede committed:
Extend application of patch to FMT

adrienbernede committed:
Fix style, add missing patch

adrienbernede committed:
Add constraint to Ascent to prevent 0.9.1 to build with umpire@2024.02.0

adrienbernede committed:
Typo

adrienbernede commented:
@trws would you agree to resume the review process ?
I updated the packages with changes from the latest release, removed the LC specific section that was cleaning gcc 4.9.3 paths (using new wrappers works).
I’d like to discuss improving rocm support in a subsequent PR.

trws commented:
Absolutely, I just got back from Japan yesterday but I'll look it over this morning.

trws commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -248,12 +260,22 @@ def initconfig_hardware_entries(self):
                 entries.append(
                     cmake_cache_string("CMAKE_CUDA_ARCHITECTURES", "{0}".format(arch_str))
                 )
+                # TODO: Additional definitions that may not be needed anymore
+                # This is an imperfect merge of the radiuss way and the "spack"
+                # way: radiuss used to define only the first arch in the list,
+                # even for CMAKE_CUDA_ARCHITECTURE. What do we want?
+                cuda_flags.append("-arch sm_{0}".format(archs[0]))
+                entries.append(cmake_cache_string("CUDA_ARCH", "sm_{0}".format(archs[0])))
```
We want the list.  On our machines where we usually do separate builds for each one, it doesn't come up as much, but in general it's better to have the list so you can compile in support for multiple compute capabilities together.  Like if you wanted to run the same binary with reasonable performance on both Pascal and Volta parts. Otherwise you have to compile for Pascal only, and lose all the special functional unit improvements.

adrienbernede replied:
Looking for how to properly specify several architectures for the `-arch` flag had me digging into the rambling discussions on the topic.
The bottom line of this quest is that I couldn’t find any mention of multiple arches passed to the `-arch` option. On the contrary, I found statements that it is in facts a shorthand for when only one arch is targeted (most common and recommended use case).
Documentation suggests that only the full syntax can be used to specify multiple architectures.

So I looked into spack, which indeed accepts several architectures to the `cuda_arch` variant:
https://github.com/spack/spack/blob/be96460ab21ebc060c11d791d914c0e204d0525b/lib/spack/spack/build_systems/cuda.py#L56
Then, there is a utility function to convert an arch list into cuda_flags:
https://github.com/spack/spack/blob/be96460ab21ebc060c11d791d914c0e204d0525b/lib/spack/spack/build_systems/cuda.py#L64

My conclusion, from your comment, is that I should be using the above static function embedded in Spack.

Refs:
https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#nvcc-examples 

Notes:
I found examples of the syntax to build fatbinaries for maximun compatibility here
https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/

adrienbernede replied:
Done, needs testing.

trws commented on var/spack/repos/builtin/packages/camp/package.py:
```diff
@@ -3,29 +3,9 @@
 from spack.package import *
 
 
-def hip_repair_options(options, spec):
-    # there is only one dir like this, but the version component is unknown
-    options.append(
-        "-DHIP_CLANG_INCLUDE_PATH="
-        + glob.glob("{}/lib/clang/*/include".format(spec["llvm-amdgpu"].prefix))[0]
-    )
-
-
-def hip_repair_cache(options, spec):
-    # there is only one dir like this, but the version component is unknown
-    options.append(
-        cmake_cache_path(
-            "HIP_CLANG_INCLUDE_PATH",
-            glob.glob("{}/lib/clang/*/include".format(spec["llvm-amdgpu"].prefix))[0],
-        )
-    )
```
Really glad to see these go, thanks!

adrienbernede replied:
You're welcome!

trws commented on var/spack/repos/builtin/packages/fmt/fmt-no-variable-initialize_10.0.0.patch:
I assume this works, but this patch looks backwards, like it's removing the initialization rather than adding it?

adrienbernede replied:
I remember having an error with the empty initialization, so not surprised to see the patch removing it, but I agree with your statement.
@adayton1 is there something to learn here ?

adayton1 replied:
From what I understand, the code in mainstream fmt is correct (`= {}` should be just fine). The patch is a workaround for what I think may be a bug in ibm-clang compilers.

adayton1 replied:
I think the patch originated from Axom.

trws replied:
Ok, as long as this is the intentional result, works for me. Thanks for the clarification.

trws commented on var/spack/repos/builtin/packages/raja/package.py:
```diff
@@ -6,33 +6,67 @@
+from .blt import llnl_link_helpers
+
+
+# Starting with 2022.03.0, the only submodule we want to fetch is tpl/desul
+# since there is no package for it. Other RAJA submodules are defined as
+# dependencies.
```
I'm not holding this up for this, but we clearly need to make a desul package available.  Adding it to my list, thanks for pointing out the lack.

trws reviewed (changes requested):
I'm pretty happy with where this is. Would like to know specifically about the patch and that todo, otherwise I think this is good to go.

rhornung67 commented on var/spack/repos/builtin/packages/raja/package.py:
```diff
@@ -123,32 +157,54 @@ class Raja(CachedCMakePackage, CudaPackage, ROCmPackage):
+    )
+
     depends_on("blt", type="build")
+    depends_on("blt@0.6.0:", type="build", when="@2024.02.0:")
```
```suggestion
    depends_on("blt@0.6.1:", type="build", when="@2024.02.0:")
```

Correct?

adrienbernede replied:
Good question. 0.6.2 is now out too. I’ll have to test whether it works with 0.6.0. Then it’s about what you want @rhornung67.

adrienbernede replied:
I see no reason not to prefer blt@0.6.1 over blt@0.6.0, so I enforced 0.6.1.
I haven’t tested 0.6.2 yet though.

adrienbernede replied:
Done.

rhornung67 replied:
We can change to BLT 0.6.2 when we update the RAJA submodule to that version.

rhornung67 commented on var/spack/repos/builtin/packages/raja/package.py:
```diff
@@ -123,32 +157,54 @@ class Raja(CachedCMakePackage, CudaPackage, ROCmPackage):
 
     variant("openmp", default=True, description="Build OpenMP backend")
-    variant("shared", default=True, description="Build Shared Libs")
+    variant("shared", default=True, description="Build shared libs")
```
Should these defaults be changed to match the defaults in RAJA CMake? Specifically, they are openmp - false, shared - false, desul - false, vectorization - false

trws replied:
Shared, desul and vectorization almost certainly.  I didn't realize the OpenMP default was false now, is that a fallback to host sequential?

rhornung67 replied:
We decided to make all parallel back-ends have the same default status.

adrienbernede replied:
Done

rhornung67 reviewed (commented):
A couple of questions about TPL versions and config defaults

adrienbernede committed:
Set CUDA_ARCH to support multiple architectures

adrienbernede committed:
Set parallel back-ends support to false by default in RAJA

adrienbernede committed:
Enforce preference of blt@0.6.1 over blt@0.6.0

rhornung67 commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -263,16 +263,10 @@ def initconfig_hardware_entries(self):
+                entries.append(cmake_cache_string("CMAKE_CUDA_ARCHITECTURES", ";".join(archs)))
+                # Additional definitions that may not be needed anymore
+                cuda_flags = cuda_flags(archs)
+                entries.append(cmake_cache_string("CUDA_ARCH", " ".join(cuda_flags)))
```
I may be wrong, but I think `CMAKE_CUDA_ARCHITECTURES` covers it and we don't need `CUDA_ARCH` anymore.

white238 replied:
correct.. `CUDA_ARCH` was never a real variable and was only referenced in other variables that then got used, for example, `CMAKE_CUDA_FLAGS`. It can be removed.

adrienbernede replied:
I want to believe you. I’ll try, but I’ll test in every RADIUSS projects. I seem to remember that it wasn’t that simple.

trws replied:
I think you're right about that for older versions of cmake, we had workarounds where they didn't quite always behave the same IIRC, something like before 3.18 or 3.21 maybe? If we're requiring what I think we are now though it should be fine. Especially with the updated BLT.

adrienbernede replied:
Works fine without CUDA_ARCH. This is done.

adrienbernede committed:
Fix Ascent +openmp needs RAJA +openmp + syntax

adrienbernede committed:
Remove CUDA_ARCH definition

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update-start-over

trws reviewed (dismissed):
Works for me.  Thanks for all the work on this @adrienbernede, good work!

adrienbernede committed:
Fix exago expects raja+shared

adrienbernede commented:
@trws I think I just fixed the failure in exago, related to a change of default for the shared variant in RAJA.
However the error is GASNet seems unrelated to this PR: GASNet does not depend on any of the modified packages, and is not derived from the CachedCMakePackage class.

adrienbernede commented:
Also, I am tempted to ask for a "rebuild everything" since changes in CachedCMakePackage are not captured well by the rebuild detection mechanism. What do you think?

trws commented:
That's probably a good idea, I'll try to kick that off.  Agree that the GasNet issue seems separate.

trws commented:
@spackbot rebuild everything

adrienbernede commented:
@trws a quite reassuring CI run :)

eugeneswalker commented:
Please incorporate these changes to resolve the OneAPI CI stack failure:
* https://github.com/spack/spack/pull/43419

adrienbernede committed:
Merge tag 'develop-2024-03-31' into llnl/radiuss-packages-update-start-over

Snapshot release 2024-03-31

eugeneswalker reviewed (approved):
CI is green!

trws commented:
Excellent!
